processing/src/test/java/org/apache/druid/query/filter/vector/VectorValueMatcherColumnProcessorFactoryTest.java

```diff
@@ -0,0 +1,277 @@
+    Assert.assertEquals(CURRENT_SIZE, booleanMatcher.getCurrentVectorSize());
+    EasyMock.verify(vectorValueSelector);
+  }
+
+  @Test
+  public void testSingleValueString()
+  {
+    IdLookup lookup = EasyMock.createMock(IdLookup.class);
+    SingleValueDimensionVectorSelector selector =
+        EasyMock.createMock(SingleValueDimensionVectorSelector.class);
+    EasyMock.expect(selector.getCurrentVectorSize()).andReturn(CURRENT_SIZE).anyTimes();
+    EasyMock.expect(selector.getMaxVectorSize()).andReturn(VECTOR_SIZE).anyTimes();
+    EasyMock.expect(selector.getValueCardinality()).andReturn(1024).anyTimes();
+    EasyMock.expect(selector.nameLookupPossibleInAdvance()).andReturn(false).anyTimes();
+    EasyMock.expect(selector.idLookup()).andReturn(lookup).anyTimes();
+    EasyMock.expect(lookup.lookupId("any value")).andReturn(1).anyTimes();
+    EasyMock.expect(lookup.lookupId("another value")).andReturn(-1).anyTimes();
+    EasyMock.replay(selector, lookup);
+
+    VectorValueMatcherFactory matcherFactory =
+        VectorValueMatcherColumnProcessorFactory.instance().makeSingleValueDimensionProcessor(selector);
+
+    Assert.assertTrue(matcherFactory instanceof SingleValueStringVectorValueMatcher);
+
+    // value exists in column nonboolean matcher
+    VectorValueMatcher matcher = matcherFactory.makeMatcher("any value");
+    Assert.assertFalse(matcher instanceof BooleanVectorValueMatcher);
+    Assert.assertEquals(VECTOR_SIZE, matcher.getMaxVectorSize());
+    Assert.assertEquals(CURRENT_SIZE, matcher.getCurrentVectorSize());
+
+    // value not exist in dictionary uses boolean matcher
+    VectorValueMatcher booleanMatcher = matcherFactory.makeMatcher("another value");
+    Assert.assertTrue(booleanMatcher instanceof BooleanVectorValueMatcher);
+    Assert.assertEquals(VECTOR_SIZE, booleanMatcher.getMaxVectorSize());
+    Assert.assertEquals(CURRENT_SIZE, booleanMatcher.getCurrentVectorSize());
+    EasyMock.verify(selector, lookup);
+  }
+
+  @Test
+  public void testSingleValueStringZeroCardinalityAlwaysBooleanMatcher()
+  {
+    // cardinality 0 has special path to always use boolean matcher
+    SingleValueDimensionVectorSelector selector =
+        EasyMock.createMock(SingleValueDimensionVectorSelector.class);
+    EasyMock.expect(selector.getCurrentVectorSize()).andReturn(CURRENT_SIZE).anyTimes();
+    EasyMock.expect(selector.getMaxVectorSize()).andReturn(VECTOR_SIZE).anyTimes();
+    EasyMock.expect(selector.getValueCardinality()).andReturn(0).anyTimes();
+    EasyMock.replay(selector);
+
+    VectorValueMatcherFactory matcherFactory =
+        VectorValueMatcherColumnProcessorFactory.instance().makeSingleValueDimensionProcessor(selector);
+
+    Assert.assertTrue(matcherFactory instanceof SingleValueStringVectorValueMatcher);
+
+    VectorValueMatcher matcher = matcherFactory.makeMatcher("any value");
+    Assert.assertTrue(matcher instanceof BooleanVectorValueMatcher);
```

Comment:
If value cardinality == 0, shouldn't it return this thing...
     ```
 // Evaluate "lookupName" and "predicate" on every row.
      return new BaseVectorValueMatcher(selector)
      {
        final VectorMatch match = VectorMatch.wrap(new int[selector.getMaxVectorSize()]);

        @Override
        public ReadableVectorMatch match(final ReadableVectorMatch mask)
        {
          final int[] vector = selector.getRowVector();
          final int[] selection = match.getSelection();

          int numRows = 0;

          for (int i = 0; i < mask.getSelectionSize(); i++) {
            final int rowNum = mask.getSelection()[i];
            if (predicate.apply(selector.lookupName(vector[rowNum]))) {
              selection[numRows++] = rowNum;
            }
          }

          match.setSelectionSize(numRows);
          assert match.isValid(mask);
          return match;
        }
      };
```

Reply:
With cardinality of 0, `ValueMatchers.toBooleanIfPossible` will return 'false' from the path from https://github.com/apache/druid/blob/17cf8ea8f2cc6a4fe8e2c6e29e3b9f066a6ccc2b/processing/src/main/java/org/apache/druid/query/filter/vector/SingleValueStringVectorValueMatcher.java#L62, which will make a false boolean matcher, so the current test is correct afaict.

Reply:
Ahh I see. Got it.